Kabanov. Task 2

kswiftie opened the pull request.

kswiftie committed:
Released all functions, fixed task1.md and task2.md

kswiftie committed:
1. Decorators functions are distributed across files.

2. Added new tests, old tests were renamed, added utils.py for tests

3. Updated tasks files

kswiftie committed:
code style fixed

kswiftie committed:
Added support for positional arguments to smart_args and tests for this

kswiftie committed:
added marks to task2.md

Krekep commented on project/decorators/smart_args.py:
```diff
@@ -0,0 +1,54 @@
+
+    argsln = len(argsinfo.args) if argsinfo.args else 0
+    defaultsln = len(argsinfo.defaults) if argsinfo.defaults else 0
+    po = argsln - defaultsln
```
po?

kswiftie replied:
Забыл сначала дать этому нормально описание. Исправил

Krekep commented on tests/test_decorators_currying.py:
```diff
@@ -0,0 +1,61 @@
+)
+def test_curried_print(function, function_arity, function_inputs, expected_result):
+    f2 = curry_explicit(function, function_arity)
+    sys.stdout = io.StringIO()  # Redirect output
```
Изящнее было бы сделать через `functools.partial`, к примеру:
```
new_output = io.StringIO()
redirected_print = partial(print, file=new_output)
redirected_print("some stuff")
# print(redirected_print.getvalue())
```

kswiftie replied:
Исправил

Krekep commented on tests/test_decorators_currying.py:
Добавьте тест, что у `print` (или другой встроенной функции произвольной арности) действительно ограничивается количество аргументов и ей нельзя передать больше аргументов, чем арность

kswiftie replied:
Добавил в существующий тест, что каррированный print по 2 аргументам от двух аргументов вернет None, то есть в него больше нельзя будет передавать аргументы. Это правильно?
Я имею в виду, что type(curry_exclipt(print, 2)(1)(2)) is None

Krekep commented on tests/test_decorators_smartargs.py:
```diff
@@ -0,0 +1,95 @@
+    ],
+)
+def test_smartargs_Isolation(function, kwargs, expected_result):
+    inp = [x for x in kwargs.values()][0]
+    function(d=inp)
+    assert inp == expected_result
```
Это можно переписать проще

kswiftie replied:
Исправил

Krekep commented on tests/test_decorators_smartargs.py:
```diff
@@ -0,0 +1,95 @@
+    assert (
+        len(set(x)) == lensetx
+        and len(set(y)) == lensety
+        and len(y) == len(y)
```
Эта проверка не несёт полезной смысловой нагрузки

kswiftie replied:
А как мне тогда проверять работу Evaluation? Я проверяю, что аргументы x, которые передаются в значение по умолчанию не Evaluated разные, а аргументы y, которые передаются в значение по умолчанию Evaluated разные. В этом ведь и смысл Evaluated, значения у будут вычисляться при каждом вызове функции

kswiftie replied:
Исправил

Krekep commented on tests/test_decorators_smartargs.py:
Не хватает тестов.
1. Смешанные позиционные и keyword-only аргументы
2. На подстановку вместо Evaluated значения при его передаче
3. На использование в функции Evaluated и Isolated вместе (как корректное использование, так и не корректное)

kswiftie replied:
Вопрос по поводу п. 3. А какое использование корректное? В Isolated передавать аргументы нельзя, а в Evaluated передается функция, которая в качестве аргумента принимает функцию без аргументов, которая что-то возвращает, Isolated под это описание не подходит

kswiftie replied:
Исправил

kswiftie replied:
Добавить оба типа в 1 функцию

kswiftie replied:
Исправил

Krekep reviewed (changes requested):
Необходимы исправления

kswiftie committed:
1. Tests for smart_args have been completely rewritten

2. Functions in utils have been fixed

3. Some fixes was made in smart_args

4. Minor improvements for currying tests

5. Files in /tasks were fixed

kswiftie commented:
# Задача 2. Декораторы

* **Дедлайн**: 17.10.2024, 23:59
* Полный балл: 15

## Задача

- [x] Частичное применение (каррирование, curry) --- это превращение функции от нескольких параметров в функцию от одного параметра,
возвращающую функцию от остальных параметров. К ней существует обратная операция --- uncarry.
В Python каррирование в таком виде затруднено из-за произвольной арности.
Напишите функцию `curry_explicit(function, arity)` и парную к ней `uncurry_explicit(function, arity)`.
- Требования/допущения:
  - [x] Если пользователь ошибся при указании арности, то бросить исключение.
  Можно [здесь](https://docs.python.org/3/tutorial/errors.html) посмотреть что это и как с этим работать.
  Здесь имеется в виду отслеживание простых ситуаций вида: пользователь указал, что функция от 1 аргумента,
  а передает 5 или же передает аргументы и указывает, что арность --- ноль.
  НЕ требуется проверки, действительно ли передана функция от трех аргументов, если передали арность = 3.
  - [x] Отрицательная арность невозможна --- бросить исключение
  - [x] При арности 0 результатом будет функция от 0 параметров: `curry_explicit(f,0)() == f()`
  - [x] При арности 1 --- функция от 1 параметра: `curry_explicit(f,1)(x) == f(x)`
  - [x] Не забудьте, что есть произвольно-арные функции, такие как print(); curry_explicit должна замораживать их арность,
  т.е. нельзя написать `curry_explicit(print,2)(1)(2)(3,4,5)`.
  В данном случае `curry_explicit(print,2)(1)(2)` вернет `None`,
  что сделает невозможным дальнейшее применение (Python сам бросит ошибку, данный кейс отдельно обрабатывать не нужно)
  - [x] Именованные аргументы поддерживать не нужно
- Пример работы:
    ```python
    f2 = curry_explicit((lambda x, y, z: f'<{x},{y}, {z}>'), 3)
    g2 = uncurry_explicit(f2, 3)
    print(f2(123)(456)(562))
    print(g2(123, 456, 562))
    ________
    <123,456,562>
    <123,456,562>
    ```
-----
- [x] Реализовать декоратор для кеширования результатов выполнения функции.
То есть необходимо сохранять результаты вычислений для разного набора аргументов функции.
Аргумент декоратора – сколько последних результатов хранить.
По умолчанию ничего не кэшируется. Необходимо поддержать как неименованные аргументы, так и именованные.
-----
- [x] Напишите декоратор `@smart_args`, который анализирует типы значений по умолчанию аргументов функции и,
в зависимости от этого, копирует и/или вычисляет их перед выполнением функции:
  - `Evaluated(func_without_args)` --- подставляет значение по умолчанию, вычисляемое в момент вызова.
  В качестве аргумента принимает функцию БЕЗ аргументов, которая что-то возвращает.
  Например, она может менять глобальный счетчик или что-то считать внутри класса.
  Во время работы необходимо вызвать эту функцию и использовать значение, которое будет получено в результате выполнения.
  - `Isolated()` --- это фиктивное значение по умолчанию; аргумент должен быть передан, но в момент передачи - скопирован (глубокая копия).
  - Также необходимо реализовать сами `Evaluated` и `Isolated`.
- Требования\допущения:
  - [x] Необходима поддержка только именованных аргументов
  - [x] Не нужно сочетать `Isolated` и `Evaluated`
  - [x] Добавьте проверки (assert) того, что пользователь случайно не вышел за эти рамки
  (использует `Isolated` и `Evaluated` в сочетании или пытается их использовать для позиционных аргументов)
- Пример работы `Isolated`:
    ```python
    @smart_args
    def check_isolation(*, d=Isolated()):
      d['a'] = 0
      return d

    no_mutable = {'a': 10}

    print(check_isolation(d=no_mutable))
    print(no_mutable)
    ________
    {'a': 0}
    {'a': 10}
    ```
- Пример работы `Evaluated` *(так как генерируется случайное число, то запуск на вашем компьютере может отличаться, но все числа в первой колонке ВСЕГДА будут одинаковыми)*:
    ```python
    import random

    def get_random_number():
       return random.randint(0,100)

    # Для x дефолтное значение всегда будет одинаковое, оно вычислится один раз
    # Для y дефолтное значение может быть разным, так как оно будет
    # вычисляться КАЖДЫЙ раз, если не передано значение y
    @smart_args
    def check_evaluation(*, x=get_random_number(), y=Evaluated(get_random_number)):
       print(x, y)


    check_evaluation()
    check_evaluation()
    check_evaluation(y=150)
    ________
    15 36
    15 66
    15 150
    ```
- [x] **(Доп., +5 баллов)** Добавить поддержку позиционных аргументов (необходимо разобраться с `inspect.getfullargspec()`).
В таком случае необходимо добавить настройку для декоратора - возможность отключить данную функциональность.

kswiftie committed:
Added 2 tests for smart_args

kswiftie committed:
Fixed code style

Krekep reviewed (approved):
Зачтено
